### PR TITLE
Add readme for deploying collector to ECS EC2

### PR DESCRIPTION
**Pull request by bjsignalfx**

### CODE REVIEW

File: cmd/otelcol/config/collector/fargate_config.yaml [hunk 1]
```diff
@@ -52,6 +52,8 @@ receivers:
     listenAddress: 0.0.0.0:9080
   smartagent/ecs-metadata:
     type: ecs-metadata
+    metadataEndpoint: "${ECS_TASK_METADATA_ENDPOINT}"
+    statsEndpoint: "${ECS_TASK_STATS_ENDPOINT}"
```
Comment (dmitryax) [inline cmd/otelcol/config/collector/fargate_config.yaml:56]: Can you please move fargate related changes to a different PR?

File: deployments/ecs/ec2/README.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
+# Amazon ECS EC2 Deployment
+Familiarity with Amazon ECS using launch type EC2 is assumed. Consult the 
+[Getting started with the Amazon ECS console using Amazon EC2](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/getting-started-ecs-ec2.html)
+for further reading.
+
+Unless stated otherwise, the
+[Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector)
+(Collector) is deployed as a **sidecar** (additional container) to ECS tasks.
```
Comment (dmitryax) [inline deployments/ecs/ec2/README.md:8]: Why it has to be sidecar for ECS on EC2? Can't we deploy it as a separate task in Daemon mode? Similar to [signalfx-agent instructions](https://github.com/signalfx/signalfx-agent/tree/main/deployments/ecs). I think it'll be much easier for user and will give them host metrics out of the box.

File: deployments/ecs/ec2/README.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
+# Amazon ECS EC2 Deployment
```
Comment (dmitryax) [inline deployments/ecs/ec2/README.md:1]: If this doc is under `deployments/ecs/ec2`, should we put fargate to `deployments/ecs/fargate`? Or move this one to `deployments/ecs`

Comment (bjsignalfx) [reply deployments/ecs/ec2/README.md:1]: Yes, I think eventually we should move fargate to `deployments/ecs/fargate`.

File: cmd/otelcol/config/collector/fargate_config.yaml (hunk 1 above)
Comment (bjsignalfx) [reply cmd/otelcol/config/collector/fargate_config.yaml:56]: Yes

File: deployments/ecs/ec2/README.md (hunk 2 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/README.md]: I changed the instructions as you suggested. I still have to add host metrics to the default config.

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      grpc:
+        endpoint: 0.0.0.0:14250
+      thrift_http:
+        endpoint: 0.0.0.0:14268
+  otlp:
+    protocols:
+      grpc:
+        endpoint: 0.0.0.0:4317
+      http:
+        endpoint: 0.0.0.0:55681
```
Comment (flands) [inline cmd/otelcol/config/collector/ecs_ec2_config.yaml:32]: Is this needed? Not sure this is enabled in any other config by default

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    traces:
+      receivers: [jaeger, otlp, zipkin, smartagent/signalfx-forwarder]
+      processors:
+        - memory_limiter
+        - batch
+        - resourcedetection
+      #        - resource/add_environment
+      exporters: [sapm, signalfx]
+    metrics:
+      receivers: [signalfx, smartagent/signalfx-forwarder, smartagent/ecs-metadata, prometheus/internal]
```
Comment (flands) [inline cmd/otelcol/config/collector/ecs_ec2_config.yaml:123]: I believe internal metrics need to be processed in a different pipeline with different configuration -- at least this is the case for helm

File: deployments/ecs/ec2/README.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    refresh_interval: 10s
+    cluster_name: 'lorem-ipsum-cluster'
+    cluster_region: 'us-west-2'
+    result_file: '/etc/ecs_sd_targets.yaml'
+    task_definitions:
+      - arn_pattern: "^arn:aws:ecs:us-west-2:906383545488:task-definition/lorem-ipsum-task:[0-9]+$"
+        metrics_ports: [9113]
+        metrics_path: /metrics
+receivers:
+  signalfx:
```
Comment (flands) [inline deployments/ecs/ec2/README.md:120]: nit: this is configured, but not enabled in the pipeline

File: deployments/ecs/ec2/splunk-otel-collector.json [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        {
+          "readOnly": true,
+          "containerPath": "/etc/passwd",
+          "sourceVolume": "etc-passwd"
+        }
+      ],
+      "workingDirectory": null,
+      "secrets": null,
+      "dockerSecurityOptions": null,
+      "memory": 1024,
```
Comment (flands) [inline deployments/ecs/ec2/splunk-otel-collector.json:55]: Isn't memory limit of collector 512 by default? Why is this 1024?

File: deployments/ecs/ec2/splunk-otel-collector.json [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      "memoryReservation": null,
+      "volumesFrom": [],
+      "stopTimeout": null,
+      "image": "quay.io/signalfx/splunk-otel-collector:0.32.0",
```
Comment (flands) [inline deployments/ecs/ec2/splunk-otel-collector.json:59]: can we use `latest` here?

**Review by flands — COMMENTED** [review]
Let's be sure to open a task to add to data setup page as well

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml (hunk 4 above)
Comment (bjsignalfx) [reply cmd/otelcol/config/collector/ecs_ec2_config.yaml:32]: Not sure if needed but this http endpoint is enabled in all default configs!

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 7 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:55]: Changed the value to 512 MiB

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 8 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json]: Ok

File: deployments/ecs/ec2/README.md (hunk 6 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/README.md:120]: Yes, I removed it.

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml (hunk 5 above)
Comment (bjsignalfx) [reply cmd/otelcol/config/collector/ecs_ec2_config.yaml:123]: Ok, I moved internal metrics to a separate pipeline.

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml (hunk 4 above)
Comment (flands) [reply cmd/otelcol/config/collector/ecs_ec2_config.yaml:32]: The port changes to 4318 in the latest release not sure if/when we should switch to that port

File: deployments/ecs/ec2/README.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+The Collector image Dockerfile is available [here](../../../cmd/otelcol/Dockerfile) and the contents of the default
+configuration file can be seen [here](../../../cmd/otelcol/config/collector/ecs_ec2_config.yaml).
+
+The configured network mode for the task is **host**. This means that **task metadata endpoint
+version 2** used by receiver `smartagent/ecs-metadata` will not be enabled by default. See
+[here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html)
+for the task metadata endpoint version that will be enabled by default for your task.
+
+Add the following to the **environment** list in the task definition JSON if **task metadata
+endpoint version 3** will be enabled for your task:
```
Comment (rmfitzpatrick) [inline deployments/ecs/ec2/README.md:30]: Not version 4 like the envvars used?

File: deployments/ecs/ec2/splunk-otel-collector.json [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+          "value": "MY_SPLUNK_REALM"
+        }
+      ],
+      "resourceRequirements": null,
+      "ulimits": null,
+      "dnsServers": null,
+      "mountPoints": [
+        {
+          "readOnly": true,
+          "containerPath": "/hostfs",
```
Comment (rmfitzpatrick) [inline deployments/ecs/ec2/splunk-otel-collector.json:38]: Why is this needed?

File: deployments/ecs/ec2/splunk-otel-collector.json [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+          "sourceVolume": "hostfs"
+        },
+        {
+          "readOnly": true,
+          "containerPath": "/var/run/docker.sock",
+          "sourceVolume": "docker-socket"
+        },
+        {
+          "readOnly": true,
+          "containerPath": "/etc/passwd",
```
Comment (rmfitzpatrick) [inline deployments/ecs/ec2/splunk-otel-collector.json:48]: Also not clear why this is needed.

File: deployments/ecs/ec2/README.md (hunk 9 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/README.md:30]: No, version 3 the endpoint env var is ECS_CONTAINER_METADATA_URI whereas for version 4 it is ECS_CONTAINER_METADATA_URI_V4.

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 10 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:38]: This supposed to be mount point for the host "/" but the volumes below are not correct!

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 11 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:48]: Mount point for /etc/passwd/ in host. https://github.com/signalfx/signalfx-agent/blob/main/deployments/ecs/signalfx-agent-task.json#L41

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml (hunk 4 above)
Comment (bjsignalfx) [reply cmd/otelcol/config/collector/ecs_ec2_config.yaml:32]: Living the port the same for now.

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 11 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json]: I removed this mount.

File: cmd/otelcol/config/collector/ecs_ec2_config.yaml [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    type: ecs-metadata
+    metadataEndpoint: "${ECS_TASK_METADATA_ENDPOINT}"
+    statsEndpoint: "${ECS_TASK_STATS_ENDPOINT}"
+    excludedImages: ${env:ECS_METADATA_EXCLUDED_IMAGES}
+
+processors:
+  batch:
+  # Enabling the memory_limiter is strongly recommended for every pipeline.
+  # Configuration is based on the amount of memory allocated to the collector.
+  # In general, the ballast should be set to 1/3 of the collector's memory, the limit
```
[the remark this answers is not in the document]
Comment (bjsignalfx) [reply cmd/otelcol/config/collector/ecs_ec2_config.yaml]: I added ballast comments as in link.

File: deployments/ecs/ec2/splunk-otel-collector.json [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+          "value": "MY_SPLUNK_REALM"
+        }
+      ],
+      "resourceRequirements": null,
+      "ulimits": null,
+      "dnsServers": null,
+      "mountPoints": [
+        {
+          "readOnly": true,
+          "containerPath": "/hostfs",
```
Comment (rmfitzpatrick) [inline deployments/ecs/ec2/splunk-otel-collector.json:38]: Is this acceptable?
```suggestion
          "containerPath": "/",
```

Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:38]: @rmfitzpatrick my immediate reaction is no because you effectively overriding the container fs with host fs! I don't see how you then get files specific to the container. @dmitryax what do you think?

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 10 above)
Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:38]: Changed to mounting host root to container /hostfs

File: deployments/ecs/ec2/splunk-otel-collector.json (hunk 13 above)
Comment (rmfitzpatrick) [reply deployments/ecs/ec2/splunk-otel-collector.json:38]: Right, you are describing potential fallout/the general issue and this is where awareness of the host metrics receiver scraper functionality and the collector's [scratch filesystem](https://github.com/signalfx/splunk-otel-collector/blob/main/cmd/otelcol/Dockerfile#L23) is required.  Using `/hostfs` won't magically resolve the problem unless there's a corresponding config change in the collector, which doesn't appear to exist for anything but the filesystem scraper.

Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:38]: @dmitryax it looks like we have to exclude host metrics unless somehow replacing the container fs with host fs is ok which I don't see how that can fly. Or, maybe there is a way to configure the root path in the collector for the scrapers to find the relevant filesystems to use.

Comment (dmitryax) [reply deployments/ecs/ec2/splunk-otel-collector.json]: Host metrics scrapers use this library which is configurable through env vars https://github.com/shirou/gopsutil#usage. Look at k8s configuration as a reference https://github.com/signalfx/splunk-otel-collector-chart/blob/main/helm-charts/splunk-otel-collector/templates/daemonset.yaml#L186-L198

Comment (bjsignalfx) [reply deployments/ecs/ec2/splunk-otel-collector.json:62]: Ok, I set the env vars that configure fs paths needed by the scraper.

**Review by rmfitzpatrick — APPROVED** [review]
